src/dune_engine/action_builder.mli

```diff
@@ -210,6 +210,58 @@ val create_file : Path.Build.t -> Action.t With_targets.t
 (** Merge a list of actions accumulating the sets of their targets. *)
 val progn : Action.t With_targets.t list -> Action.t With_targets.t
 
+module Action_desc : sig
+  (* jeremiedimino: this type correspond to a subset of [Rule.t]. We should
+     eventually share the code. *)
+  type nonrec t =
+    { context : Build_context.t option
+    ; env : Env.t option
+    ; action : Action.t
+    ; locks : Path.t list
+    ; loc : Loc.t option
+    ; dir : Path.Build.t
+          (** Directory the action is attached to. This is the directory where
+              the outcome of the action will be cached. *)
+    ; alias : Alias.Name.t option  (** For better error messages *)
+    }
+end
+
+(** Execute an action. You can think of [action t] as a convenient way of
+    declaring an anonymous build rule and depending on its outcome. While the
+    return type is [unit], the action might fail. So the outcome here is success
+    or failure. This function is commonly used for attaching tests to an alias.
+
+    Note that any dependency declared in [t] is treated as a dependency of the
+    action returned by [t], rather than the action currently being computed.
+    More precisely, in the following code:
+
+    {[
+      let+ () = Action_builder.path p1
+      and+ () =
+        Action_builder.action
+          (let+ () = Action_builder.path p2 in
+           act2)
+      in
+      act1
+    ]}
+
+    Dune assumes that:
+
+    - [act1] will read [p1]
+    - [act2] will read [p2]
+
+    When passing [--force] to Dune, these are exactly the actions that will be
+    re-executed. *)
```

Comment:
I'm confused by the difference in semantics between `action` and `action_stdout` with respect to the `--force` flag. Why not re-execute `action_stdout` too? If the action is indeed non-deterministic then it would be possible to catch this problem by rebuilding with `--force`.

Reply:
(It also makes it difficult to reason about what `--force` does: my previous mental model was pretty simple -- if you pass `--force`, all actions will be re-executed.)

Reply:
Yeah, it's probably not the right distinction. Let me change that.

Reply:
Not that I care about `--force`, but this is marked as resolved even though the comments are still saying that `action` is different from `action_stdout` in this respect. The resolution is not clear to me.

Reply:
I think I marked it resolved by accident.

Reply:
Comment is updated